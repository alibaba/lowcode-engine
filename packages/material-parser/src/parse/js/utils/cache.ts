export interface ICache {
  [name: string]: any;
}

const cache: ICache = {};

export function set(scope: string, name: string, value: any) {
  cache[scope] = cache[scope] || {};
  cache[scope][name] = value;
}

export function get(scope: string, name: string) {
  return (cache[scope] || {})[name];
}

export function has(scope: string, name: string) {
  return cache[scope] && Object.prototype.hasOwnProperty.call(cache[scope], name);
}
