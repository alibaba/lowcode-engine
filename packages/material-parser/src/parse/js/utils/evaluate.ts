const { namedTypes: t } = require('ast-types');

const { resolveToValue } = require('react-docgen').utils;

function isInfinity(path) {
  return t.Identifier.check(path.node) && path.node.name === 'Infinity';
}

function wrapValue(value, confident = true) {
  return {
    value,
    confident,
  };
}

export default function evaluate(path: any) {
  if (t.UnaryExpression.check(path.node)) {
    if (path.node.operator === 'void') {
      return wrapValue(undefined);
    }

    const argument = path.get('argument');

    const result = evaluate(argument);
    if (!result.confident) {
      return wrapValue(undefined, false);
    }

    const arg = result.value;
    if (arg === undefined) {
      return wrapValue(undefined);
    }

    switch (path.node.operator) {
      case '!':
        return wrapValue(!arg);
      case '+':
        return wrapValue(+arg);
      case '-':
        return wrapValue(-arg);
      case '~':
        return wrapValue(~arg);
      case 'typeof':
        return wrapValue(typeof arg);
    }
  }

  if (t.Identifier.check(path.node)) {
    const valuePath = resolveToValue(path);
    if (isInfinity(valuePath)) {
      return wrapValue(undefined);
    }
    return evaluate(valuePath);
  }

  if (t.Literal.check(path.node)) {
    return wrapValue(path.node.value);
  }

  if (t.ObjectExpression.check(path.node)) {
    const returnValue = {};
    path.get('properties').each((propertyPath) => {
      const { confident, value } = evaluate(propertyPath.get('value'));
      if (!confident) {
        return;
      }
      const keyPath = propertyPath.get('key');
      let key;
      if (keyPath.node.computed) {
        const result = evaluate(keyPath);
        if (!result.confident) {
          return;
        }
        key = result.value;
      } else {
        key = keyPath.node.name;
      }
      returnValue[key] = value;
    });

    return wrapValue(returnValue);
  }

  if (t.ArrayExpression.check(path.node)) {
    const value = [];
    let isValid = true;
    path.get('elements').each((x) => {
      if (!isValid) return;
      const result = evaluate(x);
      if (!result.confident) {
        isValid = false;
      } else {
        value.push(result.value);
      }
    });

    if (isValid) {
      return wrapValue(value);
    }
  }

  return wrapValue(undefined, false);
}
