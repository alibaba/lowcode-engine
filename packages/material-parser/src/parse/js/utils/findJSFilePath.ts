import fs from 'fs';

const suffixes = ['js', 'jsx'];
export default function findJSFilePath(fileBasePath: string): string {
  let filePath;
  for (const suffix of suffixes) {
    const fp = `${fileBasePath}.${suffix}`;
    if (fs.existsSync(fp)) {
      filePath = fp;
      break;
    }
  }

  return filePath;
}