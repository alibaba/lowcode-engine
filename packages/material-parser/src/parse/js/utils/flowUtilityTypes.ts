/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import { namedTypes as t } from 'ast-types';

const supportedUtilityTypes = new Set(['$Exact', '$ReadOnly']);

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
function isSupportedUtilityType(path: any) {
  if (t.GenericTypeAnnotation.check(path.node)) {
    const idPath = path.get('id');
    return !!idPath && supportedUtilityTypes.has(idPath.node.name);
  }
  return false;
}

export { isSupportedUtilityType };

function isReactUtilityType(path: any) {
  if (t.TSTypeReference.check(path.node)) {
    const objName = path.get('typeName', 'left').node.name;
    if (objName === 'React') {
      return true;
    }
  }

  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
function unwrapUtilityType(path: any) {
  while (isSupportedUtilityType(path) || isReactUtilityType(path)) {
    path = path.get('typeParameters', 'params', 0);
  }

  return path;
}

export { unwrapUtilityType };
