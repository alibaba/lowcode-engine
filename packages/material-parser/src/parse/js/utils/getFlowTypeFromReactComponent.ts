import { namedTypes as t } from 'ast-types';
import resolveGenericTypeAnnotation from './resolveGenericTypeAnnotation';

const {
  isReactComponentClass,
  isReactForwardRefCall,
  getTypeAnnotation,
  resolveToValue,
  getMemberValuePath,
} = require('react-docgen').utils;
const getTypeParameters = require('react-docgen/dist/utils/getTypeParameters').default;

function getStatelessPropsPath(componentDefinition: any) {
  const value = resolveToValue(componentDefinition);
  if (isReactForwardRefCall(value)) {
    const inner = resolveToValue(value.get('arguments', 0));
    return inner.get('params', 0);
  }
  if (t.VariableDeclarator.check(componentDefinition.parent.node)) {
    const id = componentDefinition.parent.get('id');
    if (id.node.typeAnnotation) {
      return id;
    }
  }

  return value.get('params', 0);
}

/**
 * Given an React component (stateless or class) tries to find the
 * flow type for the props. If not found or not one of the supported
 * component types returns null.
 */
export default (path: any) => {
  let typePath = null;

  if (isReactComponentClass(path)) {
    const superTypes = path.get('superTypeParameters');

    if (superTypes.value) {
      const params = superTypes.get('params');
      if (params.value.length === 3) {
        typePath = params.get(1);
      } else {
        typePath = params.get(0);
      }
    } else {
      const propsMemberPath = getMemberValuePath(path, 'props');
      if (!propsMemberPath) {
        return null;
      }

      typePath = getTypeAnnotation(propsMemberPath.parentPath);
    }

    return typePath;
  }

  const propsParam = getStatelessPropsPath(path);

  if (propsParam) {
    typePath = getTypeAnnotation(propsParam);
  }

  return typePath;
};

function applyToFlowTypeProperties(documentation: any, path: any, callback: any, typeParams?: any) {
  if (path.node.properties) {
    path.get('properties').each((propertyPath: any) => callback(propertyPath, typeParams));
  } else if (path.node.members) {
    path.get('members').each((propertyPath: any) => callback(propertyPath, typeParams));
  } else if (path.node.type === 'InterfaceDeclaration') {
    if (path.node.extends) {
      applyExtends(documentation, path, callback, typeParams);
    }

    path.get('body', 'properties').each((propertyPath: any) => callback(propertyPath, typeParams));
  } else if (path.node.type === 'TSInterfaceDeclaration') {
    if (path.node.extends) {
      applyExtends(documentation, path, callback, typeParams);
    }

    path.get('body', 'body').each((propertyPath: any) => callback(propertyPath, typeParams));
  } else if (path.node.type === 'IntersectionTypeAnnotation' || path.node.type === 'TSIntersectionType') {
    path
      .get('types')
      .each((typesPath: any) => applyToFlowTypeProperties(documentation, typesPath, callback, typeParams));
  } else if (path.node.type !== 'UnionTypeAnnotation') {
    // The react-docgen output format does not currently allow
    // for the expression of union types
    const typePath = resolveGenericTypeAnnotation(path);
    if (typePath) {
      applyToFlowTypeProperties(documentation, typePath, callback, typeParams);
    }
  }
}

function applyExtends(documentation: any, path: any, callback: any, typeParams: any) {
  path.get('extends').each((extendsPath: any) => {
    const resolvedPath = resolveGenericTypeAnnotation(extendsPath);
    if (resolvedPath) {
      if (resolvedPath.node.typeParameters && extendsPath.node.typeParameters) {
        typeParams = getTypeParameters(
          resolvedPath.get('typeParameters'),
          extendsPath.get('typeParameters'),
          typeParams,
        );
      }
      applyToFlowTypeProperties(documentation, resolvedPath, callback, typeParams);
    } else {
      const id = extendsPath.node.id || extendsPath.node.typeName || extendsPath.node.expression;
      if (id && id.type === 'Identifier') {
        documentation.addComposes(id.name);
      }
    }
  });
}

export { applyToFlowTypeProperties };
