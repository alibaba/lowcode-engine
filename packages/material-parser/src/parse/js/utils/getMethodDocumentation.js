/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import getTSType from './getTSType';

const { namedTypes: t } = require('ast-types');
const {
  resolveToValue,
  getFlowType,
  getParameterName,
  getPropertyName,
  getTypeAnnotation,
} = require('react-docgen').utils;
const { getDocblock } = require('react-docgen/dist/utils/docblock');

function getMethodFunctionExpression(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node)) {
    return resolveToValue(methodPath.get('right'));
  }
  // Otherwise this is a method/property node
  return methodPath.get('value');
}

function getMethodParamsDoc(methodPath) {
  const params = [];
  const functionExpression = getMethodFunctionExpression(methodPath);

  // Extract param flow types.
  functionExpression.get('params').each((paramPath) => {
    let type = null;
    const typePath = getTypeAnnotation(paramPath);
    if (typePath && t.Flow.check(typePath.node)) {
      type = getFlowType(typePath);
      if (t.GenericTypeAnnotation.check(typePath.node)) {
        type.alias = typePath.node.id.name;
      }
    } else if (typePath) {
      type = getTSType(typePath);
      if (t.TSTypeReference.check(typePath.node)) {
        type.alias = typePath.node.typeName.name;
      }
    }

    const param = {
      name: getParameterName(paramPath),
      optional: paramPath.node.optional,
      type,
    };

    params.push(param);
  });

  return params;
}

// Extract flow return type.
function getMethodReturnDoc(methodPath) {
  const functionExpression = getMethodFunctionExpression(methodPath);

  if (functionExpression.node.returnType) {
    const returnType = getTypeAnnotation(functionExpression.get('returnType'));
    if (returnType && t.Flow.check(returnType.node)) {
      return { type: getFlowType(returnType) };
    }
    if (returnType) {
      return { type: getTSType(returnType) };
    }
  }

  return null;
}

function getMethodModifiers(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node)) {
    return ['static'];
  }

  // Otherwise this is a method/property node

  const modifiers = [];

  if (methodPath.node.static) {
    modifiers.push('static');
  }

  if (methodPath.node.kind === 'get' || methodPath.node.kind === 'set') {
    modifiers.push(methodPath.node.kind);
  }

  const functionExpression = methodPath.get('value').node;
  if (functionExpression.generator) {
    modifiers.push('generator');
  }
  if (functionExpression.async) {
    modifiers.push('async');
  }

  return modifiers;
}

function getMethodName(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node) && t.MemberExpression.check(methodPath.node.left)) {
    const { left } = methodPath.node;
    const { property } = left;
    if (!left.computed) {
      return property.name;
    }
    if (t.Literal.check(property)) {
      return String(property.value);
    }
    return null;
  }
  return getPropertyName(methodPath);
}

function getMethodAccessibility(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node)) {
    return null;
  }

  // Otherwise this is a method/property node
  return methodPath.node.accessibility;
}

function getMethodDocblock(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node)) {
    let path = methodPath;
    do {
      path = path.parent;
    } while (path && !t.ExpressionStatement.check(path.node));
    if (path) {
      return getDocblock(path);
    }
    return null;
  }

  // Otherwise this is a method/property node
  return getDocblock(methodPath);
}

// Gets the documentation object for a component method.
// Component methods may be represented as class/object method/property nodes
// or as assignment expresions of the form `Component.foo = function() {}`
export default function getMethodDocumentation(methodPath) {
  if (getMethodAccessibility(methodPath) === 'private') {
    return null;
  }

  const name = getMethodName(methodPath);
  if (!name) return null;

  return {
    name,
    docblock: getMethodDocblock(methodPath),
    modifiers: getMethodModifiers(methodPath),
    params: getMethodParamsDoc(methodPath),
    returns: getMethodReturnDoc(methodPath),
  };
}
