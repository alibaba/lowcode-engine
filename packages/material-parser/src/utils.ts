import { pathExists, readFileSync, writeFile } from 'fs-extra';
import { isPlainObject } from 'lodash';
import safeEval from 'safe-eval';
import * as path from 'path';
import spawn from 'cross-spawn-promise';

export async function isNPMInstalled(args: {
  workDir: string;
  moduleDir: string;
  npmClient?: string;
}) {
  return pathExists(path.join(args.workDir, 'node_modules'));
}

export async function install(args: { workDir: string; moduleDir: string; npmClient?: string }) {
  if (await isNPMInstalled(args)) return;
  const { workDir, npmClient = 'tnpm' } = args;
  try {
    await spawn(npmClient, ['i'], { stdio: 'inherit', cwd: workDir } as any);
  } catch (e) {
    // TODO
  }
}

export async function installTypeScript(args: {
  workDir: string;
  moduleDir: string;
  npmClient?: string;
}) {
  if (await isNPMInstalled(args)) return;
  const { workDir, npmClient = 'tnpm' } = args;
  await spawn(npmClient, ['i', 'typescript'], { stdio: 'inherit', cwd: workDir } as any);
}

export async function installPeerDeps(args: {
  workDir: string;
  moduleDir: string;
  npmClient?: string;
}) {
  const { workDir, moduleDir, npmClient = 'tnpm' } = args;
  const modulePkgJsonPath = path.resolve(moduleDir, 'package.json');
  if (!(await pathExists(modulePkgJsonPath))) {
    return;
  }
  const pkgJsonPath = path.resolve(workDir, 'package.json');
  if (!(await pathExists(pkgJsonPath))) {
    return;
  }
  const modulePkgJson = await resolvePkgJson(modulePkgJsonPath);
  const pkgJson = await resolvePkgJson(pkgJsonPath);
  const { peerDependencies = {} } = modulePkgJson;
  pkgJson.dependencies = pkgJson.dependencies || {};
  pkgJson.dependencies = {
    ...pkgJson.dependencies,
    ...peerDependencies,
  };
  await writeFile(pkgJsonPath, JSON.stringify(pkgJson, null, 2));
  await spawn(npmClient, ['i'], { stdio: 'inherit', cwd: workDir } as any);
}

export async function installTypeModules(args: {
  workDir: string;
  moduleDir: string;
  npmClient?: string;
}) {
  const { workDir, moduleDir, npmClient = 'tnpm' } = args;
  const pkgJsonPath = path.resolve(moduleDir, 'package.json');
  if (!(await pathExists(pkgJsonPath))) {
    return;
  }
  await spawn(npmClient.replace('m', 'x'), ['typesync'], { stdio: 'inherit', cwd: workDir } as any);
}

export async function resolvePkgJson(pkgJsonPath: string): Promise<{ [k: string]: any }> {
  const content = await loadFile(pkgJsonPath);
  const json = JSON.parse(content);
  return json;
}

export function loadFile(filePath: string): string {
  const content: string | Buffer = readFileSync(filePath);
  if (typeof content === 'string') {
    return content;
  }
  return content.toString();
}

export function isPrimitive(val) {
  return !['object', 'function'].includes(typeof val) || val === null;
}

export function isEvaluable(value) {
  if (isPrimitive(value)) return true;
  if (Array.isArray(value)) {
    return value.every(isEvaluable);
  } else if (isPlainObject(value)) {
    return Object.keys(value).every(key => isEvaluable(value[key]));
  }
  return false;
}

export { safeEval };
