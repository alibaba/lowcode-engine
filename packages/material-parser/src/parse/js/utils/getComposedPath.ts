import { namedTypes as t, visit } from 'ast-types';
import fs from 'fs';
import path from 'path';
import getRoot from './getRoot';
import findJSFilePath from './findJSFilePath';

const buildParser = require('react-docgen/dist/babelParser').default;
const expressionTo = require('react-docgen/dist/utils/expressionTo');

const {
  resolveToValue,
  isExportsOrModuleAssignment,
} = require('react-docgen').utils;

export default function getComposedPropTypesPath(documentation, propName, p) {
  const composes: string[] = Array.from(documentation._composes);
  let _path = null;
  const root = getRoot(p).node;
  for (const compose of composes) {
    const composePath = findJSFilePath(path.resolve(path.dirname(root.__path), compose));
    if (!composePath) continue;

    const fileContent = fs.readFileSync(composePath, 'utf8');
    const parser = buildParser({ filename: composePath });
    const ast = parser.parse(fileContent);

    visit(ast, {
      visitAssignmentExpression(path: any) {
        // Ignore anything that is not `exports.X = ...;` or
        // `module.exports = ...;`
        if (!isExportsOrModuleAssignment(path)) {
          return false;
        }
        const arr = expressionTo.Array(path.get('left'));
        if (!(arr[0] === 'exports' && arr[1] === propName)) return false;

        // Resolve the value of the right hand side. It should resolve to a call
        // expression, something like React.createClass
        path = resolveToValue(path.get('right'));
        _path = path;
        return false;
      },
    });

    if (_path) {
      break;
    }
  }

  return _path;
}