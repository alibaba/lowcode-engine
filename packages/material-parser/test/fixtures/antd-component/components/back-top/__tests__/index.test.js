import React from 'react';
import { mount } from 'enzyme';
import { sleep } from '../../../tests/utils';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import BackTop from '..';

describe('BackTop', () => {
  mountTest(BackTop);
  rtlTest(BackTop);

  it('should scroll to top after click it', async () => {
    const wrapper = mount(<BackTop visibilityHeight={-1} />);
    const scrollToSpy = jest.spyOn(window, 'scrollTo').mockImplementation((x, y) => {
      window.scrollY = y;
      window.pageYOffset = y;
      document.documentElement.scrollTop = y;
    });
    window.scrollTo(0, 400);
    expect(document.documentElement.scrollTop).toBe(400);
    // trigger scroll manually
    wrapper.instance().handleScroll();
    await sleep();
    wrapper.find('.ant-back-top').simulate('click');
    await sleep(500);
    expect(document.documentElement.scrollTop).toBe(0);
    scrollToSpy.mockRestore();
  });

  it('support onClick', async () => {
    const onClick = jest.fn();
    const wrapper = mount(<BackTop onClick={onClick} visibilityHeight={-1} />);
    const scrollToSpy = jest.spyOn(window, 'scrollTo').mockImplementation((x, y) => {
      window.scrollY = y;
      window.pageYOffset = y;
    });
    document.dispatchEvent(new Event('scroll'));
    window.scrollTo(0, 400);
    // trigger scroll manually
    wrapper.instance().handleScroll();
    await sleep();
    wrapper.find('.ant-back-top').simulate('click');
    expect(onClick).toHaveBeenCalled();
    scrollToSpy.mockRestore();
  });

  it('should be able to update target', async () => {
    const wrapper = mount(<BackTop />);
    wrapper.instance().handleScroll = jest.fn();
    const container = document.createElement('div');
    wrapper.setProps({ target: () => container });
    expect(wrapper.instance().handleScroll).toHaveBeenLastCalledWith({
      target: container,
    });
    container.dispatchEvent(new Event('scroll'));
    expect(wrapper.instance().handleScroll).toHaveBeenLastCalledWith(
      expect.objectContaining({
        currentTarget: container,
        target: container,
      }),
    );
  });

  it('invalid target', async () => {
    const onClick = jest.fn();
    const wrapper = mount(<BackTop onClick={onClick} target={() => ({ documentElement: {} })} />);
    wrapper.find('.ant-back-top').simulate('click');
    expect(onClick).toHaveBeenCalled();
  });
});
