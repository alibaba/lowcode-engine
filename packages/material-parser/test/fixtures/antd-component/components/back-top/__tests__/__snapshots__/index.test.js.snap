// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackTop rtl render component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-back-top ant-back-top-rtl"
/>
`;
